# Módulo 5 Tarefa 1
## Base de nascidos vivos do DataSUS
O DataSUS disponibiliza diversos arquivos de dados com relação a seus segurados, conforme a [lei da transparência de informações públicas](https://www.sisgov.com/transparencia-acesso-informacao/#:~:text=A%20Lei%20da%20Transpar%C3%AAncia%20(LC,em%20um%20site%20na%20internet.). 

Essas informações podem ser obtidas pela internet [aqui](http://www2.datasus.gov.br/DATASUS/index.php?area=0901&item=1). Como o processo de obtenção desses arquivos foge um pouco do nosso escopo, deixamos o arquivo SINASC_RO_2019.csv` já como vai ser encontrado no DataSUS. O dicionário de dados está no arquivo `estrutura_sinasc_para_CD.pdf` (o nome do arquivo tal qual no portal do DataSUS).

### Nosso objetivo
Queremos deixar uma base organizada para podermos estudar a relação entre partos com risco para o bebê e algumas condições como tempo de parto, consultas de pré-natal etc.

#### Preparação da base
1. Carregue a base 'SINASC_RO_2019.csv'. Conte o número de registros e o número de registros não duplicados da base. Dica: você aprendeu um método que remove duplicados, encadeie este método com um outro método que conta o número de linhas. **Há linhas duplicadas?**  

2. Conte o número de valores *missing* por variável.  

3. Ok, no item anterior você deve ter achado pouco prático ler a informação de tantas variáveis, muitas delas nem devem ser interesantes. Então crie uma seleção dessa base somente com as colunas que interessam. São elas:
` 
['LOCNASC', 'IDADEMAE', 'ESTCIVMAE', 'ESCMAE', 'QTDFILVIVO', 
    'GESTACAO', 'GRAVIDEZ', 'CONSULTAS', 'APGAR5'] 
`
Refaça a contagem de valores *missings*.  

4. Apgar é uma *nota* que o pediatra dá ao bebê quando nasce de acordo com algumas características associadas principalmente à respiração. Apgar 1 e Apgar 5 são as notas 1 e 5 minutos do nascimento. Apgar5 será a nossa variável de interesse principal. Então remova todos os registros com Apgar5 não preenchido. Para esta seleção, conte novamente o número de linhas e o número de *missings*.  

5. observe que as variáveis `['ESTCIVMAE', 'CONSULTAS']` possuem o código `9`, que significa *ignorado*. Vamos assumir que o não preenchido é o mesmo que o código `9`.<br>
6. Substitua os valores faltantes da quantitativa (`QTDFILVIVO`) por zero.  
7. Das restantes, decida que valore te parece mais adequado (um 'não preenchido' ou um valor 'mais provável' como no item anterior) e preencha. Justifique. Lembre-se de que tratamento de dados é trabalho do cientista, e que estamos tomando decisões a todo o momento - não há necessariamente certo e errado aqui.  
8. O Apgar possui uma classificação indicando se o bebê passou por asfixia:
- Entre 8 e 10 está em uma faixa 'normal'. 
- Entre 6 e 7, significa que o recém-nascido passou por 'asfixia leve'. 
- Entre 4 e 5 significa 'asfixia moderada'.
- Entre 0 e 3 significa 'asfixia severa'.  

Crie uma categorização dessa variável com essa codificação e calcule as frequências dessa categorização.  
<br>
9. Renomeie as variáveis para que fiquem no *snake case*, ou seja, em letras minúsculas, com um *underscore* entre as palávras. Dica: repare que se você não quiser criar um *dataframe* novo, você vai precisar usar a opção `inplace = True`.

In [69]:
# 1) 
import pandas as pd
import requests

sinasc = pd.read_csv('SINASC_RO_2019.csv')
print(f'Shape: {sinasc.shape}')

sinasc.drop_duplicates().shape
print(f'Shape após drop_duplicates, {sinasc.shape}, continua o mesmo portanto não há duplicados.')

Shape: (27028, 69)
Shape após drop_duplicates, (27028, 69), continua o mesmo portanto não há duplicados.


In [25]:
# 2) seu código aqui
#Verifica o número ds missing values em cada coluna
def missing_verifier() -> dict:
    """
    Função que verifica coluna a coluna a quantidade de missing values.
    
    Argumentos:
        N/A.
    Retorno:
        n_nas: Dicionário com as colunas do DataFrame e seus respectivos números de NAs.
    """
    lista_nas = []
    colunas = []
    for i in sinasc.columns:
        lista_nas.append(sinasc[i].isna().sum())
        colunas.append(i)
    n_nas = dict(zip(colunas,lista_nas))    
    return n_nas

verificadas1 = missing_verifier()
verificadas1


{'ORIGEM': 0,
 'CODESTAB': 115,
 'CODMUNNASC': 0,
 'LOCNASC': 0,
 'IDADEMAE': 0,
 'ESTCIVMAE': 317,
 'ESCMAE': 312,
 'CODOCUPMAE': 2907,
 'QTDFILVIVO': 1573,
 'QTDFILMORT': 2098,
 'CODMUNRES': 0,
 'GESTACAO': 1232,
 'GRAVIDEZ': 79,
 'PARTO': 49,
 'CONSULTAS': 0,
 'DTNASC': 0,
 'HORANASC': 21,
 'SEXO': 4,
 'APGAR1': 96,
 'APGAR5': 103,
 'RACACOR': 647,
 'PESO': 0,
 'IDANOMAL': 591,
 'DTCADASTRO': 0,
 'CODANOMAL': 26814,
 'NUMEROLOTE': 0,
 'VERSAOSIST': 0,
 'DTRECEBIM': 0,
 'DIFDATA': 0,
 'DTRECORIGA': 27028,
 'NATURALMAE': 298,
 'CODMUNNATU': 298,
 'CODUFNATU': 298,
 'ESCMAE2010': 249,
 'SERIESCMAE': 12710,
 'DTNASCMAE': 40,
 'RACACORMAE': 661,
 'QTDGESTANT': 1212,
 'QTDPARTNOR': 1879,
 'QTDPARTCES': 1791,
 'IDADEPAI': 19421,
 'DTULTMENST': 10072,
 'SEMAGESTAC': 1238,
 'TPMETESTIM': 1238,
 'CONSPRENAT': 930,
 'MESPRENAT': 2867,
 'TPAPRESENT': 265,
 'STTRABPART': 947,
 'STCESPARTO': 747,
 'TPNASCASSI': 61,
 'TPFUNCRESP': 67,
 'TPDOCRESP': 14,
 'DTDECLARAC': 52,
 'ESCMAEAGR1': 249,
 'STDN

In [26]:
# 3) 
#Verifica o número ds missing values em cada coluna
def missing_verifier(lista_importatntes) -> dict:
    """
    Função que verifica coluna a coluna a quantidade de missing values.
    
    Argumentos:
        lista_importatntes: Lista com as colunas mais importatntes.
    Retorno:
        n_nas: Dicionário com as colunas do DataFrame e seus respectivos números de NAs.
    """
    lista_nas = []
    colunas = []
    for i in lista_importatntes:
        lista_nas.append(sinasc[i].isna().sum())
        colunas.append(i)
    n_nas = dict(zip(colunas,lista_nas))    
    return n_nas
lista = ['LOCNASC', 'IDADEMAE', 'ESTCIVMAE', 'ESCMAE', 'QTDFILVIVO', 'GESTACAO', 'GRAVIDEZ', 'CONSULTAS', 'APGAR5'] 
verificadas2 = missing_verifier(lista)
verificadas2

{'LOCNASC': 0,
 'IDADEMAE': 0,
 'ESTCIVMAE': 317,
 'ESCMAE': 312,
 'QTDFILVIVO': 1573,
 'GESTACAO': 1232,
 'GRAVIDEZ': 79,
 'CONSULTAS': 0,
 'APGAR5': 103}

In [62]:
# 4) 
#Verifica o número ds missing values em cada coluna
def missing_verifier(lista,df) -> dict:
    """
    Função que verifica coluna a coluna a quantidade de missing values.
    
    Argumentos:
        lista_importantes: Lista com as colunas mais importatntes.
    Retorno:
        n_nas: Dicionário com as colunas do DataFrame e seus respectivos números de NAs.
    """
    lista_nas = []
    colunas = []
    
    for i in lista:
        lista_nas.append(df[i].isna().sum())
        colunas.append(i)
    n_nas = dict(zip(colunas,lista_nas))
    df.dropna(subset='APGAR5', inplace=True)    
    return n_nas

lista_importantes = ['LOCNASC', 'IDADEMAE', 'ESTCIVMAE', 'ESCMAE', 'QTDFILVIVO', 'GESTACAO', 'GRAVIDEZ', 'CONSULTAS', 'APGAR5'] 

nas = sinasc['APGAR5'].isna().sum()
linhas = sinasc['APGAR5'].shape
print(f'O número de missing values é: {nas}\nO número de linhas é: {linhas[0]}')

verificadas3 = missing_verifier(lista,sinasc)

print('===============================')
nas = sinasc['APGAR5'].isna().sum()
linhas = sinasc['APGAR5'].shape
print(f'O número de missing values após dropna() é: {nas}\nO número de linhas após dropna() é: {linhas[0]}')

O número de missing values é: 103
O número de linhas é: 27028
O número de missing values após dropna() é: 0
O número de linhas após dropna() é: 26925


In [112]:
# 5) Observe que as variáveis `['ESTCIVMAE', 'CONSULTAS']` possuem o código `9`, que significa *ignorado*. Vamos assumir que o não preenchido é o mesmo que o código `9`.

colunas_9 = ['ESTCIVMAE', 'CONSULTAS']
sinasc = pd.read_csv('SINASC_RO_2019.csv')

missing_verifier(lista,sinasc)

print(f'Número de missing values: {sinasc["ESTCIVMAE"].isna().sum()}')
print(f'Valores Únicos: \n{sinasc["ESTCIVMAE"].value_counts()}')

for i in colunas_9:
    sinasc[i].fillna(value=9, inplace=True)

print(f'Número de missing values: {sinasc["ESTCIVMAE"].isna().sum()}')
print(f'Valores Únicos: \n{sinasc["ESTCIVMAE"].value_counts()}')


Número de missing values: 315
Valores Únicos: 
ESTCIVMAE
2.0    9906
1.0    8907
5.0    7145
4.0     451
9.0     144
3.0      57
Name: count, dtype: int64
Número de missing values: 0
Valores Únicos: 
ESTCIVMAE
2.0    9906
1.0    8907
5.0    7145
9.0     459
4.0     451
3.0      57
Name: count, dtype: int64


Repare que o número de ocorrências do valor 9 subiu de 144 para 451.

In [113]:
# 6) Seu código aqui 
print(f'Número de missing values: {sinasc["QTDFILVIVO"].isna().sum()}')
print(f'Valores Únicos: \n{sinasc["QTDFILVIVO"].value_counts()}\n')
sinasc['QTDFILVIVO'].fillna(value=0, inplace=True)
print(f'Número de missing values: {sinasc["QTDFILVIVO"].isna().sum()}')
print(f'Valores Únicos: \n{sinasc["QTDFILVIVO"].value_counts()}')

Número de missing values: 1566
Valores Únicos: 
QTDFILVIVO
1.0     9361
0.0     9305
2.0     4454
3.0     1362
4.0      475
5.0      212
6.0      106
7.0       41
8.0       23
9.0        8
10.0       6
12.0       3
11.0       1
30.0       1
14.0       1
Name: count, dtype: int64

Número de missing values: 0
Valores Únicos: 
QTDFILVIVO
0.0     10871
1.0      9361
2.0      4454
3.0      1362
4.0       475
5.0       212
6.0       106
7.0        41
8.0        23
9.0         8
10.0        6
12.0        3
11.0        1
30.0        1
14.0        1
Name: count, dtype: int64


In [119]:
# 7) seu código aqui
colunas_9 = ['ESTCIVMAE', 'CONSULTAS']
sinasc = pd.read_csv('SINASC_RO_2019.csv')

missing_verifier(lista,sinasc)

print(sinasc['GESTACAO'].value_counts())

sinasc['LOCNASC'].fillna(value=9, inplace=True)

media = sinasc['IDADEMAE'].mean()
sinasc['IDADEMAE'].fillna(value=media, inplace=True)

sinasc['ESCMAE'].fillna(value=9, inplace=True)
sinasc['GESTACAO'].fillna(value=9, inplace=True)
sinasc['GRAVIDEZ'].fillna(value=9, inplace=True)


sinasc['GESTACAO'].value_counts()

GESTACAO
37 a 41 semanas        22523
32 a 36 semanas         2234
42 semanas e mais        698
28 a 31 semanas          162
22 a 27 semanas           85
Menos de 22 semanas        7
Name: count, dtype: int64


GESTACAO
37 a 41 semanas        22523
32 a 36 semanas         2234
9                       1216
42 semanas e mais        698
28 a 31 semanas          162
22 a 27 semanas           85
Menos de 22 semanas        7
Name: count, dtype: int64

In [124]:
# 8) seu código aqui

#Entre 8 e 10 está em uma faixa 'normal'. 
#Entre 6 e 7, significa que o recém-nascido passou por 'asfixia leve'. 
#Entre 4 e 5 significa 'asfixia moderada'.
#Entre 0 e 3 significa 'asfixia severa'.  

sinasc.loc[(sinasc['APGAR5'] >= 8) & (sinasc['APGAR5'] <= 10), 'faixa_oms' ] = 'normal'
sinasc.loc[(sinasc['APGAR5'] >= 6) & (sinasc['APGAR5'] <= 7), 'faixa_oms' ] = 'asfixia_leve'
sinasc.loc[(sinasc['APGAR5'] >= 5) & (sinasc['APGAR5'] <= 4), 'faixa_oms' ] = 'asfixia_moderada'
sinasc.loc[(sinasc['APGAR5'] >= 0) & (sinasc['APGAR5'] <= 3), 'faixa_oms' ] = 'asfixia_severa'

sinasc['faixa_oms'].value_counts()

faixa_oms
normal            26463
asfixia_leve        320
asfixia_severa       74
nan                  68
Name: count, dtype: int64